File: src/sisyphus_util.cc
#include "sisyphus_util.h"

#include "model.pb.h"
#include "table_constants.h"
#include "parametric_processor.h"
#include <vector>
#include <cmath>
#include <cstdio>
#include <cstdlib>

namespace {
void polar_to_point(
    sisyphus::Segment_Point* point,
    const sisyphus::PolarCoordinate& polar) {
  point->set_angular_value(
      (int) (ANGULAR_STEPS_PER_REVOLUTION *
          SisyphusUtil::ClampBetween2Pi(polar.a()) / (2 * SisyphusUtil::pi)));
  point->set_linear_value((int) (LINEAR_STEPS_PER_SWEEP * polar.r()));
}

void point_to_polar(
    sisyphus::PolarCoordinate* polar,
    const sisyphus::Segment_Point& point) {
  polar->set_a((float) (point.angular_value() * 2 * SisyphusUtil::pi) / (float) ANGULAR_STEPS_PER_REVOLUTION);
  polar->set_r((float) point.linear_value() / (float) LINEAR_STEPS_PER_SWEEP);
}

sisyphus::Segment segment_from_points(
    const sisyphus::PolarCoordinate& start,
    const sisyphus::PolarCoordinate& end) {
  sisyphus::Segment segment;
  polar_to_point(segment.mutable_start(), start);
  polar_to_point(segment.mutable_end(), end);
  return segment;
}

bool segment_is_point(const sisyphus::Segment& segment) {
  return segment.start().angular_value() == segment.end().angular_value()
      && segment.start().linear_value() == segment.end().linear_value();
}

template<class Coord>
std::vector<Coord> ToVector(
    ::google::protobuf::RepeatedPtrField<Coord> coords) {
  std::vector<Coord> vec;
  for (const auto& coord : coords) {
    vec.push_back(coord);
  }
  return vec;
}

}

float SisyphusUtil::pi = 3.14159f;
float SisyphusUtil::r = .5f;

float SisyphusUtil::ClampBetween2Pi(float val) {
  float pi_2 = 2 * SisyphusUtil::pi;
  if (val < 0) {
    int cycles = (int) (-val / pi_2);
    return val + (cycles + 1) * pi_2;
  } else if (val >= pi_2) {
    int cycles = (int) (val / pi_2);
    return val - cycles * pi_2;
  }
  return val;
}

int SisyphusUtil::DiffBetweenAngles(int angle1, int angle2) {
  int diff = angle1 - angle2;
  int diff_magnitude = abs(diff);
  return diff_magnitude > (ANGULAR_STEPS_PER_REVOLUTION / 2)
      ? (ANGULAR_STEPS_PER_REVOLUTION - diff_magnitude) * (diff < 0 ? 1 : -1)
      : diff;
}

sisyphus::PolarCoordinate SisyphusUtil::PolarFromCartesian(
    const sisyphus::CartesianCoordinate& coordinate) {
  float x = coordinate.x();
  float y = coordinate.y();
  float r = sqrt(x * x + y * y);
  float a = atan2(y, x);
  sisyphus::PolarCoordinate new_coordinate;
  new_coordinate.set_r(r);
  new_coordinate.set_a(a);
  return new_coordinate;
}

sisyphus::Pattern SisyphusUtil::PatternFromPolarCoordinates(
    const std::vector<sisyphus::PolarCoordinate>& coordinates) {
  sisyphus::Pattern p;
  if (coordinates.empty()) {
    return p;
  }
  if (coordinates.size() == 1) {
    *p.add_path_segment() =
        segment_from_points(coordinates[0], coordinates[0]);
    return p;
  }
  sisyphus::PolarCoordinate prev = coordinates[0];
  for (int i = 1; i < coordinates.size(); i++) {
    sisyphus::PolarCoordinate current = coordinates[i];
    sisyphus::Segment segment = segment_from_points(prev, current);
    if (!segment_is_point(segment)) {
      *p.add_path_segment() = segment;
    }
    prev = current;
  }
  if (p.path_segment().empty()) {
    *p.add_path_segment() = segment_from_points(prev, prev);
  }
  return p;
}

sisyphus::Pattern SisyphusUtil::PatternFromCartesianCoordinates(
    const std::vector<sisyphus::CartesianCoordinate>& coordinates) {
  std::vector<sisyphus::PolarCoordinate> polar;
  for (const sisyphus::CartesianCoordinate& coordinate : coordinates) {
    polar.push_back(SisyphusUtil::PolarFromCartesian(coordinate));
  }
  return SisyphusUtil::PatternFromPolarCoordinates(polar);
}

sisyphus::Pattern SisyphusUtil::StoredPatternToPattern(
    const sisyphus::StoredPattern& pattern) {
  if (pattern.has_cartesian()) {
    return SisyphusUtil::PatternFromCartesianCoordinates(ToVector(pattern.cartesian().coordinate()));
  } else if (pattern.has_polar()) {
    return SisyphusUtil::PatternFromPolarCoordinates(ToVector(pattern.polar().coordinate()));
  } else if (pattern.has_parametric()) {
    return SisyphusUtil::PatternFromCartesianCoordinates(
        ParametricProcessor::GetCoordinatesFromParametricFunction(pattern.parametric()));
  }
  return sisyphus::Pattern();
}

sisyphus::StoredPattern SisyphusUtil::PatternToPolarStoredPattern(
    const sisyphus::Pattern& pattern) {
  sisyphus::StoredPattern stored_pattern;
  if (pattern.path_segment_size() == 0) {
      return stored_pattern;
  }

  sisyphus::PolarCoordinate* coord =
      stored_pattern.mutable_polar()->add_coordinate();
  point_to_polar(coord, pattern.path_segment(0).start());
  for (const auto& segment : pattern.path_segment()) {
    coord = stored_pattern.mutable_polar()->add_coordinate();
    point_to_polar(coord, segment.end());
  }

  return stored_pattern;
}

File: node/sisyphus_util.js
const Model = require('./gen/model_pb');
const {ANGULAR_STEPS_PER_REVOLUTION, LINEAR_RANGE_STEPS} = require('./table_constants');
const {getCoordinatesFromParametricFunction} = require('./parametric_processor');

function polarToPoint(polar) {
  const point = new Model.Segment.Point();
  point.setAngularValue(Math.floor(ANGULAR_STEPS_PER_REVOLUTION * ClampBetween2Pi(polar.getA()) / (2 * Math.PI)));
  point.setLinearValue(Math.floor(LINEAR_RANGE_STEPS * polar.getR()));
  return point;
}

function pointToPolar(point) {
  const polar = new Model.PolarCoordinate();
  polar.setA((point.getAngularValue() * 2 * Math.PI) / ANGULAR_STEPS_PER_REVOLUTION);
  polar.setR(point.getLinearValue() / LINEAR_RANGE_STEPS);
  return polar;
}

function segmentFromPoints(start, end) {
  const segment = Model.Segment();
  segment.setStart(polarToPoint(start));
  segment.setEnd(polarToPoint(end));
  return segment;
}

function segmentIsPoint(segment) {
  return segment.getStart().getAngularValue() == segment.getEnd().getAngularValue()
      && segment.getStart().getLinearValue() == segment.getEnd().getLinearValue();
}

function clampBetween2Pi(val) {
  const pi2 = 2 * Math.PI;
  if (val < 0) {
    const cycles = Math.floor(-val / pi2);
    return val + (cycles + 1) * pi2;
  } else if (val >= pi2) {
    const cycles = Math.floor(val / pi2);
    return val - cycles * pi2;
  }
  return val;
}

function diffBetweenAngles(angle1, angle2) {
  const diff = angle1 - angle2;
  const diffMagnitude = Math.abs(diff);
  return diffMagnitude > (ANGULAR_STEPS_PER_REVOLUTION / 2)
      ? (ANGULAR_STEPS_PER_REVOLUTION - diffMagnitude) * (diff < 0 ? 1 : -1)
      : diff;
}

function polarFromCartesian(coordinate) {
  const x = coordinate.getX();
  const y = coordinate.getY();
  const r = Math.sqrt(x * x + y * y);
  const a = Math.atan2(y, x);
  const newCoordinate = Model.PolarCoordinate();
  newCoordinate.setR(r);
  newCoordinate.setA(a);
  return newCoordinate;
}

function patternFromPolarCoordinates(coordinates) {
  const p = new Model.Pattern();
  if (coordinates.length == 0) {
    return p;
  }
  if (coordinates.length == 1) {
    p.addPathSegment(segmentFromPoints(coordinates[0], coordinates[0]));
    return p;
  }
  let prev = coordinates[0];
  for (let i = 1; i < coordinates.size(); i++) {
    const current = coordinates[i];
    const segment = segmentFromPoints(prev, current);
    if (!segmentIsPoint(segment)) {
      p.addPathSegment(segment);
    }
    prev = current;
  }
  if (p.getPathSegmentList().length == 0) {
    p.addPathSegment(segmentFromPoints(prev, prev));
  }
  return p;
}

function patternFromCartesianCoordinates(coordinates) {
  const polar = [];
  for (const coordinate of coordinates) {
    polar.push(polarFromCartesian(coordinate));
  }
  return patternFromPolarCoordinates(polar);
}

function storedPatternToPattern(pattern) {
  if (pattern.hasCartesian()) {
    return patternFromCartesianCoordinates(pattern.getCartesian().getCoordinateList());
  } else if (pattern.hasPolar()) {
    return patternFromPolarCoordinates(pattern.getPolar().getCoordinateList());
  } else if (pattern.hasParametric()) {
    return patternFromCartesianCoordinates(
      getCoordinatesFromParametricFunction(pattern.getParametric()));
  }
  return new Model.Pattern();
}

function patternToPolarStoredPattern(pattern) {
  const storedPattern = new Model.StoredPattern();
  if (pattern.getPathSegmentList().length == 0) {
    return storedPattern;
  }

  const polarCoords = [];
  polarCoords.push(pointToPolar(pattern.getPathSegmentList()[0].getStart()));

  for (const segment of pattern.getPathSegmentList()) {
    polarCoords.push(pointToPolar(segment.end()));
  }
  storedPattern.setPolar(polarCoords);
  return storedPattern;
}

module.exports = {
  clampBetween2Pi,
  diffBetweenAngles,
  polarFromCartesian,
  patternFromPolarCoordinates,
  patternFromCartesianCoordinates,
  storedPatternToPattern,
  patternToPolarStoredPattern
};